runtime/src/comm/ofi/README.md

```diff
@@ -0,0 +1,418 @@
+for executeOn transactions for on-statement bodies, memory transfers
+between unregistered memory, atomic operations which cannot be done
+natively, and several other things.
+
+The comm layer has two libfabric tools it can use to determine when
+operations are in some sense "done" and to impose order on them with
+respect to other operations:
+
+- _Completion levels_ adjust what effects of an operation must be
+  visible before libfabric delivers an event indicating its completion
+  to the initiator.
+
+- _Message ordering_ settings specify how operations may or may not be
+  reordered with respect to each other by libfabric, the provider, or
+  the network fabric.  Note that "message" in this context is a general
+  term referring to all kinds of communication operations, not just
+  those of the SEND/RECV message interface.
+
+The completion level and message ordering settings are specified for the
+entire libfabric domain, that is, the overarching interface to the
+network, rather than individually on each transaction.  (Libfabric does
+allow for some variation in completion level settings on individual
+transactions, but the comm layer does not currently make use of that.)
+
+#### Completion Levels
+
+Libfabric's completion level specifies what state an operation has to
+reach before an event is placed in the transmit endpoint's associated
+completion queue to indicate that the operation is "done".  The comm
+layer can currently work with either the *transmit-complete* or the
+*delivery-complete* completion levels.
+
+Transmit-complete, for the reliable endpoints used by comm=ofi, means
+that the operation has arrived at the target node and is no longer
+dependent on the network fabric or local (initiator-side) resources.  It
+does not say anything about the state of the operation at the target
+node.  In particular, it does not say that any of the intended effects
+of the operation, such as memory updates, are visible.
+
+Delivery-complete extends this to say that the operation's effects are
+visible.  For messages this means that the message data has been placed
+in the user buffer at the target node, although it does not say whether
+code outside of libfabric running on the target node has attended to the
+message.  For RMA it means that the receiving data location, on the
+target node for PUTs and native non-fetching atomics, or on the local
+node for GETs and native fetching atomics, has been updated.
+
+For operations that return data to the initiator, such as RMA GETs and
+native atomic-fetch, the source endpoint is also considered a
+destination endpoint and the default completion level is
+delivery-complete.  The default for other operations is determined by
+the provider.
+
+**_Note (Bug):_** The comm layer assumes that the default completion
+  level for all providers is transmit-complete.  This is a bug.  A
+  provider could certainly choose some other default.
```

Comment:
Can you say more about this? (responding to this comment is fine, I'm just not sure I'm familiar with this bug)

Reply:
Yeah, after I wrote that I began wondering if it really was a bug.  Providers are free to choose any default completion level they want for operations that don't deliver a local result, such as writes, which is what this is talking about.  As discussed in the paragraph just preceding this, the required completion level for operations that do deliver a local result, such as reads, is delivery-complete.  I think actually that the things the comm layer does to ensure MCM compliance in the absence of delivery-complete for writes and such is sufficient no matter what the provider's default is.  So what I should do is to rephrase this section to say that the comm layer can work with either delivery-complete or whatever the provider's default is, if it's less.  I'll do that.